### DISCUSSION TITLE
Side-by-side images

### DISCUSSION THREAD

**Discussion by janosh** [Witiko/markdown#302] [Q&A] [answered]
Is there an example of how to use HTML or LaTeX `subfigure` for rendering two images side-by-side? Needs to work on Overleaf which uses `Package: markdown 2022/05/31 v2.15.2` (unless there's a way to automatically update `markdown` on Overleaf?).

**Comment by Witiko** [comment]
> Is there an example of how to use HTML or LaTeX subfigure for rendering two images side-by-side?

I don't think there is such an example. However, there are multiple ways in which you could do that. Some of them require no programming effort whatsoever but will lead to less readable markdown code stuffed with formatting commands:

1. Enable the [`hybrid`][3] option and just type LaTeX code for the subfigures in your markdown document. This is the least principled solution that takes away many of the advantages of using markdown in the first place, such as readability and portability.
2. Enable the [`rawAttribute`][4] and [`fencedCode`][5] options and type LaTeX code for the subfigures between ```` ``` {=tex} ```` and ```` ``` ````. This is more principled than using the `hybrid` option, leading to higher readability and portability (raw attributes are supported by other tools such as Pandoc), but the LaTeX code is still visual clutter.
3. Enable the [`contentBlocks`][6] option, save the LaTeX code for the subfigures to separate file `images.tex`, and type `/images.tex` surrounded by blank lines in your markdown document. This one keeps your markdown code clean, but spreads your document across multiple files, making it more difficult to navigate. Nevertheless, I like to use this solution in my documents.

Other solutions will allow you to type nice high-level markdown code but require programming effort:

4. Enable the [`fencedDivs`][7] option, type markdown images between ```` ::: {.subfigure} ```` and ```` ::: ````, and have a programmer prepare TeX code that will cause subfigures of appropriate size to be rendered. The general approach to styling markdown attributes such as `.subfigure` is described in the recent TUGboat article [*Attributes for Markdown*][7].
5. Type your subfigures in any way you like and have a programmer prepare a syntax extension for the Markdown package, as described in the recent TUGboat article [*Markdown 2.17.1: What’s new, what’s next?*][9], section 2.2 (User-defined syntax extensions).

Although I can give you simple examples for points 4 and 5, I find that the solutions are often specific to the needs of the publication and best solved on a case-by-case basis. If you'd like to hire a programmer, see my [Fiverr][8] and [Github Sponsors][10] profiles, or seek out other talent on our [Matrix][11] and [Discord][12] servers.

> unless there's a way to automatically update markdown on Overleaf?

You can download [the latest release][1] of the Markdown package and install it on Overleaf as described in [the user manual][2], subsection Portable Installation.

 [1]: https://github.com/Witiko/markdown/releases/latest
 [2]: https://witiko.github.io/markdown/#installation
 [3]: https://witiko.github.io/markdown/#option-hybrid
 [4]: https://witiko.github.io/markdown/#option-rawattribute
 [5]: https://witiko.github.io/markdown/#option-fencedcode
 [6]: https://witiko.github.io/markdown/#option-contentblocks
 [7]: https://www.overleaf.com/read/dshtsnnmtshs
 [8]: https://www.fiverr.com/witiko/
 [9]: https://www.overleaf.com/read/pgwrhhskmgfm
 [10]: https://github.com/sponsors/Witiko?frequency=one-time
 [11]: https://matrix.to/#/#witiko-markdown:matrix.org
 [12]: https://discord.gg/8xJsPghzSH

**Reply by janosh** [reply, marked as answer]
Thanks for the very thorough answer. This is great! 🥇 

I had tried option 1 already but encountered errors trying to have a full-width figure in a 2-col doc. The figure appears twice and while it spans both columns, it overlaps with the text.

**Error**
```txt
<to be read again> 
                   \protect 
l.3 ...caption{Caption place holder} \end{figure}}
                                                  \markdownRendererInterbloc...
The control sequence marked <to be read again> should
not appear between \csname and \endcsname.

! Missing \endcsname inserted.
<to be read again> 
                   \protect 
l.3 ...caption{Caption place holder} \end{figure}}
                                                  \markdownRendererInterbloc...
The control sequence marked <to be read again> should
not appear between \csname and \endcsname.
```

![Screenshot 2023-05-15 at 20 26 45](https://github.com/Witiko/markdown/assets/30958850/0cc8dd0a-5a76-4af2-8121-0ffbd13b2517)


```tex
## Introduction

\begin{figure*}[t!]
    \centering
    \begin{subfigure}[t]{0.5\textwidth}
        \centering
        \includegraphics[width=\linewidth]{figures/foo.pdf}
        \caption{Lorem ipsum}
    \end{subfigure}%
    ~ 
    \begin{subfigure}[t]{0.5\textwidth}
        \centering
        \includegraphics[width=\linewidth]{figures/bar.pdf}
        \caption{Lorem ipsum, lorem ipsum,Lorem ipsum, lorem ipsum,Lorem ipsum}
    \end{subfigure}
    \caption{Caption place holder}
\end{figure*}
```

**Reply by janosh** [reply]
If I remove the `*` from `figure*`, the errors disappear, the figure is only rendered once but the text still overlaps.

**Reply by Witiko** [reply]
It would seem that the `*` in `figure*` is understood as markdown emphasis. You may need to escape it as `\*` (see also section *Peek Under the Hood* in the README.md document of the Markdown package). Alternatively, please try options 2 and 3, neither will cause your TeX code to be interpreted as markdown by accident.

There seems to be another part to your problem regarding the overlapping text but that seems difficult to debug without knowing your document class and having a minimal working example. Perhaps fixing your first problem will also fix your second problem.

**Reply by janosh** [reply]
Thanks so much! Escaping the `*` solved both problems. I had also tried [rawAttribute](https://witiko.github.io/markdown/#option-rawattribute) yesterday which required manually updating `markdown` on Overleaf since the TexLive 2022 distro doesn't contain this feature yet. Btw, what is the minimum viable set of files for a local installation? `markdown.tex`, `markdown.sty` and `markdown.lua`?

**Reply by Witiko** [reply]
Happy to have helped. If you want to parse YAML, you will also need `markdown-tinyyaml.lua`. Note that `markdown.tex` must be located in a directory named `markdown/` due to #74.

**Reply by janosh** [reply]
Do you what I might be doing wrong here? I copied the [`linkAttributes` example](https://witiko.github.io/markdown/#option-linkAttributes) into Overleaf and compiled the document using LuaLatex and the newest release of the markdown package but the option in curly brackets have no effect.:

![Screenshot 2023-05-22 at 11 51 48](https://github.com/Witiko/markdown/assets/30958850/c20385e3-5659-44d5-ad5a-e96f323100e8)

**Reply by Witiko** [reply]
Not really sure. I would suspect that the markdown package has not been properly installed, i.e. the distro version is still loaded.

See [my Overleaf project](https://www.overleaf.com/2731328641kwcmdhgtsdkc), which uses the same code as you do and gives the expected output:

 ![scrot](https://github.com/Witiko/markdown/assets/603082/656b6866-40ac-4fe1-b21b-b5cfa8154109)

Notice the files in the left side tab: Can you see any difference to your project?

**Reply by janosh** [reply]
Oh I didn't realize only the markdown.tex file belongs inside the `markdown/` folder. I had `.sty` and `.lua` in there as well. Moving them out of that folder to root level fixes the issue.

**Reply by Witiko** [reply]
It is so that users can have their own document named `markdown.tex` without the package breaking. You are less likely to have your own document named `markdown.sty`, etc. Without this insight, the choice may seem somewhat arbitrary and confusing. Please, let me know if you had some idea how to make the user manual clearer about the `markdown/` folder.

**Reply by janosh** [reply]
Once Overleaf updates to TexLive 2023, much fewer users will need to do a manual installation so this problem will at least partially solve itself given time. But I think putting the expected folder structure of a manual installation would be of help in the docs:

```txt
./markdown/markdown.tex
./markdown.lua
./markdown.sty
```

**Reply by Witiko** [reply]
As long as the package is actively developed and Overleaf lags behind with its TeX Live updates as much as it does, there will always be good reasons to manually update the package. Thanks for the suggestion, I added it in d13084b3.

**Reply by janosh** [reply]
Is there a way to use [linkAttributes](https://witiko.github.io/markdown/#option-linkAttributes) to get inline styles? The custom renderer I'm currently using is plucked straight from the docs:

```tex
\ExplSyntaxOn
\markdownSetup{
  renderers = {
    imageAttributeContextBegin = {
      \group_begin:
      \markdownSetup{
        renderers = {
          attributeKeyValue = {
            \setkeys
              { Gin }
              { { ##1 } = { ##2 } }
          },
        },
      }
    },
    imageAttributeContextEnd = {
      \group_end:
    },
  },
}
\ExplSyntaxOff
```

Maybe there's a way to extend the above TeX code to have this

```md
![fig:plot-name](figs/plot-name.pdf){width=.4\linewidth style="display: inline-block;"}
```

render as an inline image? Or similarly, have a flag in Markdown

```md
![fig:plot-name](figs/plot-name.pdf){width=.4\linewidth inline}
```

that's picked up by the custom renderer to make the image inline and be able to position multiple side-by-side.

Come to think of it, I guess that's what you we're suggesting as option 7 above?

**Reply by Witiko** [reply]
You could have a classname `.inline` atached to your images:

``` tex
![fig:plot-name](figs/plot-name.pdf){width=.4\linewidth .inline}
```

You can imagine having a boolean variable set to false at `imageAttributeContextBegin`. You would set the variable to true when you encounter `attributeClassName` with `#1 = inline` before `imageAttributeContextEnd`. Then, you would react to the boolean variable in `image` by doing whatever it is that you'd like to do.

 [1]: https://witiko.github.io/markdown/#image-renderer

Here is how the `image` renderer is defined in LaTeX by default (taken straight out of `markdown.sty`):

![image renderer definition for latex](https://github.com/Witiko/markdown/assets/603082/a86e136c-5b71-459e-a2eb-1220f85aad6b)

Regardless, this is a programming exercise rather than something the package would support out-of-box.
